internal/venice-client-common/src/main/java/com/linkedin/venice/utils/SparseConcurrentList.java

```diff
@@ -110,55 +118,117 @@ public void forEach(Consumer<? super E> itemConsumer) {
     }
   }
 
+  /**
+   * N.B.: The intent of having a separate function for this, and to not override the behavior of {@link #size()} is
+   * that if the code does a for loop using the size, we want them to be able to get every element, e.g.:
+   *
+   * <code>
+   *   for (int i = 0; i < list.size(); i++) {
+   *     doSomething(list.get(i));
+   *   }
+   * </code>
+   *
+   * An alternative is to call {@link #values()} which filters out nulls, e.g.:
+   *
+   * <code>
+   *   for (E element: list.values()) {
+   *     doSomething(element);
+   *   }
+   * </code>
+   *
+   * @return the number of non-null elements in this list, from a cached counter (updated at mutation time).
+   */
+  public int nonNullSize() {
+    return this.nonNullSize;
+  }
+
+  /**
+   * N.B.: If the list is only populated with null entries, this function will return true.
+   *
+   * @return a boolean indicating whether the list is empty
+   */
+  @Override
+  public boolean isEmpty() {
+    return nonNullSize() == 0;
+  }
+
   // Boilerplate code just to add synchronization to mutation operations:
 
+  @Override
   public synchronized boolean add(E e) {
+    handleSizeDuringMutation(null, e);
     return super.add(e);
   }
 
+  @Override
   public synchronized void add(int index, E element) {
+    handleSizeDuringMutation(null, element);
     super.add(index, element);
   }
 
+  @Override
   public synchronized boolean removeAll(Collection<?> c) {
-    return super.removeAll(c);
+    boolean modified = super.removeAll(c);
+    this.nonNullSize = values().size();
+    return modified;
   }
 
+  @Override
   public synchronized boolean retainAll(Collection<?> c) {
-    return super.retainAll(c);
+    boolean modified = super.retainAll(c);
+    this.nonNullSize = values().size();
+    return modified;
   }
 
+  @Override
   public synchronized int addAllAbsent(Collection<? extends E> c) {
     if (c == null) {
       return 0;
     }
-    return super.addAllAbsent(c);
+    int numberAdded = super.addAllAbsent(c);
+    this.nonNullSize = values().size();
+    return numberAdded;
   }
 
+  @Override
   public synchronized void clear() {
+    this.nonNullSize = 0;
     super.clear();
   }
 
+  @Override
   public synchronized boolean addAll(Collection<? extends E> c) {
-    return super.addAll(c);
+    boolean modified = super.addAll(c);
```

Comment:
Are we assuming addAll will always add non-null values?

Reply:
No we're not assuming that... that's why I'm doing this in the line below:

```
this.nonNullSize = values().size();
```

That will iterate through the entire list to figure out the number of non-null entries.

BTW, the various operations listed under `// Boilerplate code just to add synchronization to mutation operations:` are not used today... they're just added for completeness sake.

Reply:
I see, `size()` does iterate on non-null elements.  Good to have complete implementation of this util class, in case in future if someone else plans to use all the methods.